dask_saturn/core.py

```diff
@@ -100,66 +86,106 @@ def scheduler_address(self):
+            return {"workers": {}}
         url = urljoin(self.cluster_url, "scheduler_info")
         response = requests.get(url, headers=HEADERS)
         if not response.ok:
-            if self._get_pod_status() in ["error", "closed", "stopped"]:
+            self._refresh_status()
+            if self.status in ["error", "closed", "stopped"]:
                 for pc in self.periodic_callbacks.values():
                     pc.stop()
+            if self.status == "error":
                 raise ValueError("Cluster is not running.")
-            raise ValueError(response.reason)
+            return {"workers": {}}
         return response.json()
 
-    def _start(
-        self,
+    def __enter__(self):
+        self.status = "starting"
+        self.sync(self._start)
+        assert self.status in ["running", "ready"]
+        return self
+
+    async def _start(self):
+        """Start a cluster"""
+        expBackoff = ExpBackoff(wait_timeout=self.scheduler_service_wait_timeout)
+
+        while self.status == "starting":
+            print(f"Starting cluster. Status: {self.status}")
+            await expBackoff.wait()
+            if self.cluster_url is not None:
+                self._refresh_status()
+
+        if self.status in ["running", "ready"]:
+            print(f"Cluster is {self.status}")
+            return
+        if self.status == "closed":
+            raise ValueError("Cluster is closed")
+
+        self.status = "starting"
+        print(f"Starting cluster. Status: {self.status}")
+
+        cluster_config = {
+            "n_workers": self.n_workers,
+            "worker_size": self.worker_size,
+            "scheduler_size": self.scheduler_size,
+            "nprocs": self.nprocs,
+            "nthreads": self.nthreads,
+        }
+
+        url = urljoin(BASE_URL, "api/dask_clusters")
+        response = requests.post(url, data=json.dumps(cluster_config), headers=HEADERS)
+        if not response.ok:
+            raise ValueError(response.reason)
+        data = response.json()
+
+        self.status = data["status"]
+        if self.status == "error":
+            raise ValueError(" ".join(data["errors"]))
+
+        self.cluster_url = f"{url}/{data['id']}/"
+        self._dashboard_link = data["dashboard_address"]
+        self._scheduler_address = data["scheduler_address"]
+        self._refresh_status()
+
+    @classmethod
+    def reset(
+        cls,
         n_workers=None,
         worker_size=None,
         scheduler_size=None,
         nprocs=None,
         nthreads=None,
         scheduler_service_wait_timeout=DEFAULT_WAIT_TIMEOUT_SECONDS,
     ):
-        """Start a cluster that has already been defined for the project."""
-        url = urljoin(BASE_URL, "api/dask_clusters")
-        self.cluster_url = None
+        """Return a SaturnCluster
 
+        Destroy existing Dask cluster attached to the Jupyter Notebook or
+        Custom Deployment and recreate it with the given configuration.
```

Comment:
I think the "attached to the Jupyter Notebook or Custom Deployment" should be dropped in favor of "Saturn resource" or something like that. Otherwise this will be out of date the next time we add a different type of "thing" to attach Dask to, or if/when we change the name from "Custom Deployment" to "Deployment"

Reply:
:+1: